tardis/energy_input/gamma_ray_interactions.py

```diff
@@ -0,0 +1,160 @@
+import copy
+import numpy as np
+from astropy.coordinates import cartesian_to_spherical
+
+from tardis.energy_input.util import (
+    GXPacket,
+    GXPacketStatus,
+    kappa_calculation,
+    euler_rodrigues,
+    compton_theta_distribution,
+    get_random_theta_gamma_ray,
+    get_random_phi_gamma_ray,
+    normalize,
+    get_perpendicular_vector,
+)
+
+
+def get_compton_angle(energy):
+    """
+    Computes the compton angle from the Klein-Nishina equation.
+
+    Parameters
+    ----------
+    energy : float
+        Packet energy
+
+    Returns
+    -------
+    compton_angle : float
+        Compton scattering angle
+    lost_energy : float
+        Energy lost based on angle
+    new_energy : float
+        Packet energy
+    """
+    theta_angles, theta_distribution = compton_theta_distribution(energy)
+
+    z = np.random.random()
+
+    # get Compton scattering angle
+    compton_angle = np.interp(z, theta_angles, theta_distribution)
+    # Energy calculations
+    new_energy = energy / (
+        1.0 + kappa_calculation(energy) * (1.0 - np.cos(compton_angle))
+    )
+    lost_energy = energy - new_energy
+
+    return compton_angle, lost_energy, new_energy
+
+
+def compton_scatter(gxpacket, compton_angle):
+    """
+    Changes the direction of the gamma-ray by the Compton scattering angle
+
+    Parameters
+    ----------
+    gxpacket : GXPacket object
+    compton_angle : float
+
+    Returns
+    -------
+    float64
+        Packet theta direction
+    float64
+        Packet phi direction
+    """
+    # transform original direction vector to cartesian coordinates
+    original_direction = normalize(gxpacket.direction.cartesian_coords)
+    # compute an arbitrary perpendicular vector to the original direction
+    orthogonal_vector = get_perpendicular_vector(original_direction)
+    # determine a random vector with compton_angle to the original direction
+    new_vector = normalize(
+        np.dot(
+            euler_rodrigues(compton_angle, orthogonal_vector),
+            original_direction,
+        )
+    )
+
+    # draw a random angle from [0,2pi]
+    phi = 2.0 * np.pi * np.random.random()
+    # rotate the vector with compton_angle around the original direction
+    final_compton_scattered_vector = normalize(
+        np.dot(euler_rodrigues(phi, original_direction), new_vector)
+    )
+    # transform the cartesian coordinates to spherical coordinates
+    r_final, theta_final, phi_final = cartesian_to_spherical(
+        final_compton_scattered_vector[0],
+        final_compton_scattered_vector[1],
+        final_compton_scattered_vector[2],
+    )
+
+    return theta_final.value + 0.5 * np.pi, phi_final.value
```

Comment:
why do we always add 90 degrees

Reply:
Because of the astropy spherical/cartesian conversion axis choice